### PR TITLE
Stop generating once <eos> token is triggered

### LINKED ISSUE #130: Terminate early at EOS when performing inference
Our inference script (generate.py) and instruct.py (if #128 lands) currently generate text up to the max length. However, the model can generate an EOS token any time. We should stop the sequence and truncate it. In #128 this truncation was done after generating the full sequence, but we could already early stop during [the for loop](https://github.com/Lightning-AI/lit-llama/blob/094fa95523dc3f157aa357c8d1beaee6377f4459/generate.py#L42).

### PR DESCRIPTION
**Pull request by timothylimyl**
Hi there, 

I think we should just stop generating any new tokens once the <eos> token is found. The user can specify to trigger this feature by providing the <eos> token id to the `generate()` function via the argument `eos_id`.  We only need this feature for instruction tuned models, so I just added the `eos_id` argument into lora and adapter generate function respectively.

Have a good day!

Fixes #130

### CODE REVIEW

**Review by lantiga — APPROVED** [review]
Looks great, thanks for the contribution @timothylimyl !

**Comment by lantiga** [comment]
Wait, talked too soon, there are failures in the tests :-)

File: generate.py [hunk 1]
```diff
@@ -57,9 +59,13 @@ def generate(
 
         probs = torch.nn.functional.softmax(logits, dim=-1)
         idx_next = torch.multinomial(probs, num_samples=1)
-
+        
+        # if <eos> token is triggered, return the output (stop generation)
+        if idx_next==eos_id:
```
Comment (carmocca) [inline generate.py:64]: This is batched so this wont work. See my comment at https://github.com/Lightning-AI/lit-llama/issues/130#issuecomment-1505564783

**Comment by lantiga** [comment]
Weird, I can't repro the failure locally

**Comment by carmocca** [comment]
@lantiga The command to run is `pytest tests/test_generate.py::test_generate[2] -v`. Maybe you missed checking out this branch

File: generate.py (hunk 1 above)
Comment (lantiga) [reply generate.py:64]: Ah, of course you’re right.
this speeds up generation, so I’d vouch for following this approach

**Comment by lantiga** [comment]
Possible, I had multiple repos / terminals open when I gave it a shot

**Comment by timothylimyl** [comment]
Sorry about that, I was not aware that we are aiming to generate multiple batches of prompts. In all the generation scripts, the prompt is of type `str` rather than `list[str]`. Please let me know what fixes can I do,  I'm excited to contribute.

**Comment by awaelchli** [comment]
We can also drop the batch support, we don't use it anywhere in our demo scripts. It is just there as an artifact because we adapted the code from nanoGPT. IMO we should do it because it would simplify the code and a batched version of the function can always be added if needed.

**Comment by carmocca** [comment]
If we remove it, then https://github.com/Lightning-AI/lit-llama/issues/49 should be closed

**Comment by timothylimyl** [comment]
I reckon that the fixes be done here first in `tokenizer.py` if we want to support batch generation:

```
    def encode(
        self,
        string: str,
        bos: bool = True,
        eos: bool = False,
        max_length: int = -1,
        pad: bool = False,
        device: Optional[torch.device] = None
    ) -> torch.Tensor:
        tokens = self.processor.encode(string)
        if bos:
            tokens = [self.bos_id] + tokens
        if eos:
            tokens = tokens + [self.eos_id]
        if max_length > 0:
            tokens = tokens[:max_length]
        if pad and len(tokens) < max_length:
            tokens += [self.pad_id] * (max_length - len(tokens))

        return torch.tensor(tokens, dtype=torch.int, device=device)
```

I think we can take in a list of strings as default, if the length is one then add a batch dimension, if the length is above one then we loop to concatenate the `tokens` on the first dim (batch dim).  

Something along these lines (quick pseudocode):
```
    output_tokens_tensor = []
    for string in list_of_prompts:
        tokens = self.processor.encode(string)
        if bos:
            tokens = [self.bos_id] + tokens
        if eos:
            tokens = tokens + [self.eos_id]
        if max_length > 0:
            tokens = tokens[:max_length]
        if pad and len(tokens) < max_length:
            tokens += [self.pad_id] * (max_length - len(tokens))
        
        output_tokens_tensor.append(token)
    
    output_tokens_tensor = torch.stack(output_tokens_tensor)
    return output_tokens_tensor.to(device)

 
```

In `generate.py`, we can then make it:
```
    encoded_prompt = tokenizer.encode(prompt, bos=True, eos=False, device=fabric.device)
    encoded_prompt = encoded_prompt[None, :]  # add batch dimension
```
to 
```
    encoded_prompt = tokenizer.encode(prompt, bos=True, eos=False, device=fabric.device)
```

Then I will fix the generate function to deal with batches.  Let me know if you think this is worth looking into.

**Comment by timothylimyl** [comment]
> We can also drop the batch support, we don't use it anywhere in our demo scripts. It is just there as an artifact because we adapted the code from nanoGPT. IMO we should do it because it would simplify the code and a batched version of the function can always be added if needed.

I can help out with this in the future for the next PR?

**Comment by timothylimyl** [comment]
Okay, so are we finalising going with the current approach of single batch generation first? Anything to fix in this PR?

**Comment by timothylimyl** [comment]
> @timothylimyl I suggest that you remove batch support in a separate PR. When that's merged, the changes here will work

Hi @carmocca , a little confused here on what changes to make (this PR is meant to work for the current version of a single user prompt).

**Comment by carmocca** [comment]
@timothylimyl There is a test for batched generation (the one failing!), and the function implements a `B` dimension which should be removed (also docstrings etc)

My proposal is that you do this in a separate PR

**Comment by carmocca** [comment]
@timothylimyl I opened #162 to unblock this

**Review by carmocca — APPROVED** [review]

**Comment by timothylimyl** [comment]
> @timothylimyl I opened #162 to unblock this

awesome, thank you!

**Review by lantiga — APPROVED** [review]
Stamped! Thank you @timothylimyl!

**Review by awaelchli — COMMENTED** [review]
Needs #166 
Otherwise LGTM
